pkg/kubelet/volumemanager/reconciler/reconstruct_new_test.go

```diff
@@ -0,0 +1,261 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconciler
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/volume"
+	volumetesting "k8s.io/kubernetes/pkg/volume/testing"
+	"k8s.io/kubernetes/pkg/volume/util"
+)
+
+func TestReconstructVolumes(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SELinuxMountReadWriteOncePod, true)()
+
+	tests := []struct {
+		name                                string
+		volumePaths                         []string
+		postSyncStatCallback                func(rcInstance *reconciler, fakePlugin *volumetesting.FakeVolumePlugin) error
+		expectedVolumesNeedReportedInUse    []string
+		expectedVolumesNeedDevicePath       []string
+		expectedVolumesFailedReconstruction []string
+		verifyFunc                          func(rcInstance *reconciler, fakePlugin *volumetesting.FakeVolumePlugin) error
+	}{
+		{
+			name: "when two pods are using same volume and both are deleted",
+			volumePaths: []string{
+				path.Join("pod1", "volumes", "fake-plugin", "pvc-abcdef"),
+				path.Join("pod2", "volumes", "fake-plugin", "pvc-abcdef"),
+			},
+			expectedVolumesNeedReportedInUse:    []string{"fake-plugin/pvc-abcdef", "fake-plugin/pvc-abcdef"},
+			expectedVolumesNeedDevicePath:       []string{"fake-plugin/pvc-abcdef", "fake-plugin/pvc-abcdef"},
+			expectedVolumesFailedReconstruction: []string{},
+			verifyFunc: func(rcInstance *reconciler, fakePlugin *volumetesting.FakeVolumePlugin) error {
+				mountedPods := rcInstance.actualStateOfWorld.GetMountedVolumes()
+				if len(mountedPods) != 0 {
+					return fmt.Errorf("expected 0 certain pods in asw got %d", len(mountedPods))
+				}
+				allPods := rcInstance.actualStateOfWorld.GetAllMountedVolumes()
+				if len(allPods) != 2 {
+					return fmt.Errorf("expected 2 uncertain pods in asw got %d", len(allPods))
+				}
+				volumes := rcInstance.actualStateOfWorld.GetPossiblyMountedVolumesForPod("pod1")
+				if len(volumes) != 1 {
+					return fmt.Errorf("expected 1 uncertain volume in asw got %d", len(volumes))
+				}
+				return nil
+			},
+		},
+		{
+			name: "when reconstruction fails for a volume, volumes should be cleaned up",
+			volumePaths: []string{
+				path.Join("pod1", "volumes", "missing-plugin", "pvc-abcdef"),
+			},
+			expectedVolumesNeedReportedInUse:    []string{},
+			expectedVolumesNeedDevicePath:       []string{},
+			expectedVolumesFailedReconstruction: []string{"pvc-abcdef"},
+		},
+	}
```

Comment:
We are missing a test that was included in fix - https://github.com/kubernetes/kubernetes/pull/110670 . For this PR, I know the test will be somewhat simpler, because all we need to check is if volume is in DSOW, whether it makes it into ASOW.

Reply:
I'm not sure I understand, the new `reconstructVolumes` does not check DSOW at all, because it's not yet populated. Volume reconstruction adds all reconstructed volumes to ASOW and failed to `volumesFailedReconstruction`. `TestCleanOrphanVolumes` then checks a failed volume is / is not cleaned by `cleanOrphanVolumes` if it is not / is in DSOW.

Reply:
I was looking for a test that triggers following scenarios:

1. Volume gets picked during reconsturction and added to ASOW (verify that)
2. Mount fails and verify that volume is still in uncertain state in ASOW.

Reply:
1. Volume gets picked during reconstuction and added to ASOW 
2. DSOW gets populated
3. reconciler tries to mount, mount fails
4. check that it's still uncertain in ASOW?

I can try to do that. But I don't see how it's better than two already existing tests:

* 1. Volume gets picked during reconstuction and added to ASOW 
  2. check that it's uncertain in ASOW


* 1. volume is uncertain in ASOW and it is in DSOW
  2. reconciler tries to mount and fails
  3. check that it's still uncertain in ASOW

Reply:
Now I see! Adding unit tests that checks `asw.foundDuringReconstruction` is set for reconstructed volumes.

Reply:
New unit tests added + I actually fixed regression of #110670. Good catch!